## Analyse provided data

1. How many locations are there in the data set?
2. What are the five locations with the highest number of records?
3. How many service settings, context types, national categories, and appointment statuses are there?

### Import Pandas and prepare DataFrames

In [2]:
# Import packages with standard conventions
import numpy as np
import pandas as pd

ad = pd.read_csv('actual_duration.csv')
ar = pd.read_csv('appointments_regional.csv')
nc = pd.read_excel('national_categories.xlsx')

# View the DataFrames.
print(ad.shape)
print(ad.columns)

print(ar.shape)
print(ar.columns)

print(nc.shape)
print(nc.columns)

(137793, 8)
Index(['sub_icb_location_code', 'sub_icb_location_ons_code',
       'sub_icb_location_name', 'icb_ons_code', 'region_ons_code',
       'appointment_date', 'actual_duration', 'count_of_appointments'],
      dtype='object')
(596821, 7)
Index(['icb_ons_code', 'appointment_month', 'appointment_status', 'hcp_type',
       'appointment_mode', 'time_between_book_and_appointment',
       'count_of_appointments'],
      dtype='object')
(817394, 8)
Index(['appointment_date', 'icb_ons_code', 'sub_icb_location_name',
       'service_setting', 'context_type', 'national_category',
       'count_of_appointments', 'appointment_month'],
      dtype='object')


### Determine descriptive statistics

In [3]:
ad.describe()

count_of_appointments
count          137793.000000
mean             1219.080011
std              1546.902956
min                 1.000000
25%               194.000000
50%               696.000000
75%              1621.000000
max             15400.000000

In [4]:
ar.describe()

count_of_appointments
count          596821.000000
mean             1244.601857
std              5856.887042
min                 1.000000
25%                 7.000000
50%                47.000000
75%               308.000000
max            211265.000000

In [5]:
nc.describe()

count_of_appointments
count          817394.000000
mean              362.183684
std              1084.576600
min                 1.000000
25%                 7.000000
50%                25.000000
75%               128.000000
max             16590.000000

### How many locations are there in the data set?

In [50]:
# Count how many locations using the nc DatFrame
locations = nc['sub_icb_location_name'].value_counts()

print(locations, '\n')
print(f"There are {len(locations)} locations in total.")

NHS North West London ICB - W2U3Z              13007
NHS Kent and Medway ICB - 91Q                  12637
NHS Devon ICB - 15N                            12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS North East London ICB - A3A8R              11837
                                               ...  
NHS North East and North Cumbria ICB - 00N      4210
NHS Lancashire and South Cumbria ICB - 02G      4169
NHS Cheshire and Merseyside ICB - 01V           3496
NHS Cheshire and Merseyside ICB - 01T           3242
NHS Greater Manchester ICB - 00V                2170
Name: sub_icb_location_name, Length: 106, dtype: int64 

There are 106 locations in total.


### What are the five locations with the highest number of records?

In [12]:
# Count which locations have the highest number of records
nc['sub_icb_location_name'].value_counts().to_frame()

sub_icb_location_name
NHS North West London ICB - W2U3Z                            13007
NHS Kent and Medway ICB - 91Q                                12637
NHS Devon ICB - 15N                                          12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V                  12171
NHS North East London ICB - A3A8R                            11837
...                                                            ...
NHS North East and North Cumbria ICB - 00N                    4210
NHS Lancashire and South Cumbria ICB - 02G                    4169
NHS Cheshire and Merseyside ICB - 01V                         3496
NHS Cheshire and Merseyside ICB - 01T                         3242
NHS Greater Manchester ICB - 00V                              2170

[106 rows x 1 columns]

There are 106 locations.

Top 5 most number of records are:

NHS North West London ICB - W2U3Z	13007\
NHS Kent and Medway ICB - 91Q	12637\
NHS Devon ICB - 15N	12526\
NHS Hampshire and Isle Of Wight ICB - D9Y0V	12171\
NHS North East London ICB - A3A8R	11837

**Which areas have the most number of appointments:**

In [29]:
# Determine the most number of appointments among the locations
loc_sum = locations.groupby(['sub_icb_location_name'])['count_of_appointments']\
.agg(sum).sort_values(ascending=False)

# View the DataFrame
print(loc_sum)

sub_icb_location_name
NHS North West London ICB - W2U3Z               12142390
NHS North East London ICB - A3A8R                9588891
NHS Kent and Medway ICB - 91Q                    9286167
NHS Hampshire and Isle Of Wight ICB - D9Y0V      8288102
NHS South East London ICB - 72Q                  7850170
                                                  ...   
NHS Cheshire and Merseyside ICB - 01V             641149
NHS Nottingham and Nottinghamshire ICB - 02Q      639660
NHS Greater Manchester ICB - 00V                  639211
NHS Cheshire and Merseyside ICB - 01T             606606
NHS Lancashire and South Cumbria ICB - 02G        554694
Name: count_of_appointments, Length: 106, dtype: int64


### How many service settings, context types, national categories, and appointment statuses are there?

**Service settings:**

In [35]:
# Count how many service settings are there
nc['service_setting'].value_counts().to_frame()

service_setting
General Practice                    359274
Primary Care Network                183790
Other                               138789
Extended Access Provision           108122
Unmapped                             27419

In [51]:
# Count how many service settings are there
serv_set = nc['service_setting'].value_counts().to_frame()

print(serv_set, '\n')
print(f"There are {len(serv_set)} service settings used.")

                           service_setting
General Practice                    359274
Primary Care Network                183790
Other                               138789
Extended Access Provision           108122
Unmapped                             27419 

There are 5 service settings used.


**Context types:**

In [36]:
# Count how many context types are there
nc['context_type'].value_counts().to_frame()

context_type
Care Related Encounter        700481
Inconsistent Mapping           89494
Unmapped                       27419

In [52]:
# Count how many context types are there
con_types = nc['context_type'].value_counts().to_frame()

print(con_types, '\n')
print(f"There are {len(con_types)} context types used.")

                        context_type
Care Related Encounter        700481
Inconsistent Mapping           89494
Unmapped                       27419 

There are 3 context types used.


**National categories:**

In [37]:
# Count how many national categories are there
nc['national_category'].value_counts().to_frame()

national_category
Inconsistent Mapping                                            89494
General Consultation Routine                                    89329
General Consultation Acute                                      84874
Planned Clinics                                                 76429
Clinical Triage                                                 74539
Planned Clinical Procedure                                      59631
Structured Medication Review                                    44467
Service provided by organisation external to th...              43095
Home Visit                                                      41850
Unplanned Clinical Activity                                     40415
Patient contact during Care Home Round                          28795
Unmapped                                                        27419
Care Home Visit                                                 26644
Social Prescribing Service                                      26492
Care Home Needs Assessment & Personalised Care ...              23505
Non-contractual chargeable work                                 20896
Walk-in                                                         14179
Group Consultation and Group Education                           5341

In [53]:
# Count how many national categories are there
nat_cat = nc['national_category'].value_counts().to_frame()

print(nat_cat, '\n')
print(f"There are {len(nat_cat)} national categories used.")

                                                    national_category
Inconsistent Mapping                                            89494
General Consultation Routine                                    89329
General Consultation Acute                                      84874
Planned Clinics                                                 76429
Clinical Triage                                                 74539
Planned Clinical Procedure                                      59631
Structured Medication Review                                    44467
Service provided by organisation external to th...              43095
Home Visit                                                      41850
Unplanned Clinical Activity                                     40415
Patient contact during Care Home Round                          28795
Unmapped                                                        27419
Care Home Visit                                                 26644
Social Prescribing S

**Appointment statuses:**

In [38]:
# Count how many appointment statuses are there
ar['appointment_status'].value_counts().to_frame()

appointment_status
Attended              232137
Unknown               201324
DNA                   163360

In [54]:
# Count how many appointment statuses are there
app_stat = ar['appointment_status'].value_counts().to_frame()

print(app_stat, '\n')
print(f"There are {len(app_stat)} appointment statuses used.")

          appointment_status
Attended              232137
Unknown               201324
DNA                   163360 

There are 3 appointment statuses used.


In [55]:
app_stat_subset = ar[['icb_ons_code', 'appointment_status', 'count_of_appointments']]
print(app_stat_subset)

       icb_ons_code appointment_status  count_of_appointments
0         E54000034           Attended                   8107
1         E54000034           Attended                   6791
2         E54000034           Attended                  20686
3         E54000034           Attended                   4268
4         E54000034           Attended                  11971
...             ...                ...                    ...
596816    E54000050            Unknown                     21
596817    E54000050            Unknown                      8
596818    E54000050            Unknown                     28
596819    E54000050            Unknown                     17
596820    E54000050            Unknown                     10

[596821 rows x 3 columns]


In [58]:
# Determine the total count of status to understand opportunities in data
app_stat_sum = app_stat_subset.groupby(['appointment_status'])['count_of_appointments']\
.agg(sum).sort_values(ascending=False)

# View the DataFrame
print(app_stat_sum)

appointment_status
Attended    677755876
Unknown      34137416
DNA          30911233
Name: count_of_appointments, dtype: int64


### Are there any insights and trends identified while determining the results?

As already indicated there are a lot of quality issues in the data and that is apparent here.  However, looking at the content itself:
1. On apppointment status, which we will dwell on in this project:
    - 91% of the regional appointments show that patients have attended their appointments.  there is 5% of Did not Attend (DNA) and 4% Unknown that we have an opportunity to look into further.  This is where the unnecessary costs are.

2. On national categories and context types:
    - From a data sanitation and streamlining perspective, it would be beneficial if the national categories are standardised so that time spent and future pritoritisation can be reviewed more accurately.
    - Inconsistent mapping is at 5th position while Unmapped items is currently 7th (in volume) out of 18 categories.  This is a sizeable 13% of transactions that we cannot trace at the moment.